back/src/forms/mutations/save-form.ts

```diff
@@ -0,0 +1,73 @@
+import { Context } from "../../types";
+import { flattenObjectForDb, unflattenObjectFromDb } from "../form-converter";
+import { DomainError, ErrorCode } from "../../common/errors";
+import { getReadableId } from "../readable-id";
+import {
+  Form,
+  FormUpdateInput,
+  FormCreateInput
+} from "../../generated/prisma-client";
+import { getUserCompanies } from "../../companies/helper";
+
+export async function saveForm(_, { formInput }, context: Context) {
+  const userId = context.user.id;
+
+  const { id, ...formContent } = formInput;
+  const form = flattenObjectForDb(formContent);
+
+  await checkThatUserIsPartOftheForm(userId, { ...form, id }, context);
+
+  if (id) {
+    const updatedForm = await context.prisma.updateForm({
+      where: { id },
+      data: {
+        ...(form as FormUpdateInput),
+        appendix2Forms: { connect: formContent.appendix2Forms }
+      }
+    });
+
+    return unflattenObjectFromDb(updatedForm);
+  }
+
+  const newForm = await context.prisma.createForm({
+    ...(form as FormCreateInput),
+    appendix2Forms: { connect: formContent.appendix2Forms },
+    readableId: await getReadableId(context),
+    owner: { connect: { id: userId } }
+  });
+
+  return unflattenObjectFromDb(newForm);
+}
+
+const formSiretsGetter = (form: Partial<Form>) => [
+  form.emitterCompanySiret,
+  form.traderCompanySiret,
+  form.recipientCompanySiret,
+  form.transporterCompanySiret
+];
+
+async function checkThatUserIsPartOftheForm(
+  userId: string,
+  form: Partial<Form>,
+  context: Context
+) {
+  const isEdition = form.id != null;
+  const formSirets = formSiretsGetter(form);
+  const hasPartialFormInput = formSirets.some(siret => siret == null);
+
+  if (isEdition && hasPartialFormInput) {
+    const savedForm = await context.prisma.form({ id: form.id });
+    const savedFormSirets = formSiretsGetter(savedForm);
+    formSirets.push(...savedFormSirets);
+  }
+
+  const userCompanies = await getUserCompanies(userId);
+  const userSirets = userCompanies.map(c => c.siret);
+
+  if (!formSirets.some(siret => userSirets.includes(siret))) {
+    throw new DomainError(
+      "Vous ne pouvez pas créer ou modifier un bordereau sur lequel votre entreprise n'apparait pas.",
+      ErrorCode.BAD_USER_INPUT
```

Comment:
Il me semble que c'est plus une erreur `Not authorized`. Est-ce qu'on ne pourrait pas l'implémenter sous forme de permission graphql-shield d'ailleurs ?

Reply:
Bonne question, j'ai hésité au moment de dev...
J'avais en tete le cas ou on veut enregistrer un brouillon incomplet sans avoir saisi ses informations entreprise. On tomberai alors selon moi plutot dans du `BAD_USER_INPUT`.
Pour le resteon penche surement plus vers du `NOT_AUTHORIZED`.

Comme vous le sentez ! Si on le passe en `NOT_AUTHORIZED` on le sort en shield ouaip.

Reply:
Mmmh en effet je n'avais pas pensé à ce cas de figure, effectivement ça ressemblerait plutôt à un `BAS_USER_INPUT` dans ce cas. Dans le doute laisse comme ça, ça t'évite du boulot de refactoring

Reply:
J'ai essayé de reposer le truc pour mieux y réfléchir:
- l'utilisateur a le droit de créer ou modifier ce bordereau (il est bien authentifié, ou présent sur le bordereau, sinon il se fait virer par le shield)
- il nous envoie un `formInput` pour créer ou modifier ce bordereau
- on analyse cet input pour prendre une décision => si jamais il n'apparait pas dans l'input on le bloque. Il se retrouve alors dans un statut `NOT_AUTHORIZED`, mais causé par un `BAD_USER_INPUT`

J'ai donc le sentiment que l'erreur de base est un `BAD_USER_INPUT` dans tous les cas.

Dans les shields on analyse souvent un état en DB (es tu admin ? transporteur ? authentifié?), alors que la on interdit un changement d'état induit par l'input (création ou modification de bordereau).

Ca te semble cohérent ?